### ISSUE TITLE
Alternative implementation of Stream.filter

### ISSUE THREAD

**Issue by asarkar** [underscoreio/advanced-scala#260]
Chapter 3 implements `Stream.filter` as follows:
```scala
def filter(pred: A => Boolean): Stream[A] = {
    val self = this
    new Stream[A] {
      def head: A = {
        def loop(stream: Stream[A]): A =
          if pred(stream.head) then stream.head
          else loop(stream.tail)
          
        loop(self)
      }
      
      def tail: Stream[A] = {
        def loop(stream: Stream[A]): Stream[A] =
          if pred(stream.head) then stream.tail
          else loop(stream.tail)
          
        loop(self)
      }
    }
  }
  ```
  
Whereas, I implemented it as follows:
```scala
def filter(p: A => Boolean): Stream[A] =
  lazy val self = if p(head) then this else tail.filter(p)
  new Stream[A]:
    def head: A         = self.head
    def tail: Stream[A] = self.tail.filter(p)
```

Is my implementation functionally equivalent, or is there something I'm missing?

**Comment by asarkar** [comment]
It appears the difference is that when given an infinite stream, the book's implementation being tail-recursive runs forever, while my implementation runs throws a `StackOverflowError`. For example:
```
val ones: Stream[Int] =
  new Stream[Int]:
    def head: Int = 1
    def tail: Stream[Int] = ones

ones.filter(_ % 2 == 0).take(5) // never-ending
```

**Comment by noelwelsh** [comment]
There is an error in the book's code which is addressed in PR #253 

You are correct about the difference between your implementation and mine.
